### PR TITLE
Dask compatibility

### LINKED ISSUE #61: [v0.2] Large memory overhead in cutout.pv
Different topic:
I failed to run `cutout.pv(...)` for the last days and found the problem to be insufficient memory (the script ran slowly and crashed at one point in time without a tractable error).
I solved it with `windows="D"`, which requires less data to be loaded into memory at the same time.

@coroa : What do you think about making `windows="D"` default for the convert operations?

_Originally posted by @euronion in https://github.com/PyPSA/atlite/pull/20#issuecomment-617057624_

### PR DESCRIPTION
**Pull request by FabianHofmann**
This PR introduces a full dask compatible version. It builds on branch v0.2, but has further improvements on performance and structure.

closes #61
solves #48

# User effective changes 

* Choosing a module (e.g. `module='sarah'`), will only retrieve variables which are provided by the module. 
* Choosing more than one module is supported now (e.g. `module=['sarah', 'era5']`) 
* `layout` in conversion functions must be a `xr.DataArray`
* `atlite.Cutout` takes the argument `chunks`, which serves as a default for chunking
* atlite now default to chunking when loading or preparing a cutout 
* all conversion functions are automatically using dask
* `class Window` is removed due to replacement by dask

# Main changes under the hook

* datasets module only handle features that the underlying data provides. So, `sarah.py` is strongly reduced to sarah data handling functionality only. 
* the `get_data`  in all modules returns data in a standard format, always `xr.Dataset` with chunkes `dask.arrays`. No delayed dask objects. Using `xr.Dataset` with `dask.arrays` performs better and allows to directly apply standard `xarray` functions on the dataset "skeleton" (as setting attributes, renaming, etc).    
* loading `features` are coordinated by the `data.py` module, it looks which of the modules in `Cuotut.module` provides the requested feature and gets it from there. If more than one module provides a feature (or to be more precise: a variable, e.g. `influx_diffuse`) it gets it from the first  module in the module list.  

# Remaining issues 
- [x] clarify small deviations in `heat_demand` calculation

### CODE REVIEW

File: atlite/gis.py [hunk 1]
```diff
@@ -14,30 +14,65 @@
...
+    """
+    x = slice(*sorted([x.start, x.stop]))
+    y = slice(*sorted([y.start, y.stop]))
+
+    ds = xr.Dataset({'x': np.arange(-180, 180, dx),
+                     'y': np.arange(-90, 90, dy),
+                     'time': pd.date_range(start="1979", end="now", freq=dt)})
+    ds = ds.assign_coords(lon=ds.coords['x'], lat=ds.coords['y'])
+    ds = ds.sel(x=x, y=y, time=time)
+    return ds
```
Comment (coroa) [inline atlite/gis.py:67]: Should we add the concept of a projection into this `get_coords` function? Think rotpole transform for cordex for example.

I'd propose to have this as a separate issue, though!

File: atlite/gis.py [hunk 2]
```diff
@@ -72,7 +109,7 @@ def reproject_shapes(shapes, p1, p2):
 
     if isinstance(p1, RotProj):
         if p2 == 'latlong':
-            reproject_points = lambda x,y: p1(x, y, inverse=True)
+            def reproject_points(x, y): return p1(x, y, inverse=True)
```
Comment (coroa) [inline atlite/gis.py:112]:
```suggestion
            def reproject_points(x, y):
                return p1(x, y, inverse=True)
```

File: atlite/utils.py [hunk 3]
```diff
@@ -8,39 +8,45 @@
...
+            pgb.widgets.Timer(),
+            ' ',
+            pgb.widgets.ETA()]
         if not prefix.endswith(": "):
             prefix = prefix.strip() + ": "
         maybe_progressbar = pgb.ProgressBar(prefix=prefix, widgets=widgets,
                                             max_value=max_value)
     else:
-        maybe_progressbar = lambda x: x
+        def maybe_progressbar(x): return x
```
Comment (coroa) [inline atlite/utils.py:43]:
```suggestion
        def maybe_progressbar(x):
            return x
```

File: atlite/utils.py [hunk 4]
```diff
@@ -67,21 +73,27 @@ def migrate_from_cutout_directory(old_cutout_dir):
...
-                             "with the command above!")
+            logger.exception(
+                "Automatic migration failed. Re-create the cutout "
+                "with the command above!")
             raise
 
-    data.attrs['prepared_features'] = list(sys.modules['atlite.datasets.' + data.attrs["module"]].features)
+    data = maybe_swap_spatial_dims(data)
+    data.attrs['prepared_features'] = list(
+        sys.modules['atlite.datasets.' + data.attrs["module"]].features)
```
Comment (coroa) [inline atlite/utils.py:84]:
```suggestion
    data.attrs['prepared_features'] = list(
        atlite.datasets.modules[data.attrs["module"]].features)
```

File: atlite/utils.py [hunk 5]
```diff
@@ -67,21 +73,27 @@ def migrate_from_cutout_directory(old_cutout_dir):
...
+
+    path = Path(path).with_suffix(".nc")
+    logger.info(f"Writing cutout data to {path}. When done, load it again using"
+                f"\n\n\tatlite.Cutout('{path}')")
+    data.to_netcdf(path)
 
-    return data
```
Comment (coroa) [inline atlite/utils.py:79]: Why no return?

You could return the opened cutout instead!

**Comment by FabianHofmann** [comment]
Thanks for reviewing so fast! I will comment where there are question or things are unclear, otherwise I'll commit the suggestions or add them remotely.

File: atlite/data.py [hunk 6]
```diff
@@ -146,117 +92,127 @@ def __iter__(self):
...
+        ds = ds[missing_vars.values]
+        ds = ds.assign_attrs(cutout.data.attrs)
 
-        # Write to a temporary file in the same directory first and then move back,
-        # because we might still want to load data from the original file in the process
-        directory, filename = os.path.split(str(cutout.path))
-        fd, target = mkstemp(suffix=filename, dir=directory)
-        os.close(fd)
+        # netCDF4 does not permit boolean values. Convert to str to preserve information
+        ds.attrs.update({k:v if not isinstance(v, bool) else str(v) for k,v in ds.attrs.items()})
```
[the remark this answers is not in the document]
Comment (coroa) [reply atlite/data.py:201]: You could just store them as an int, instead!

if int(False):
    print('Not reached')
else:
    print('0 is actually fals-ish')

File: atlite/utils.py (hunk 5 above)
Comment (FabianHofmann) [reply atlite/utils.py:79]: right

**Comment by FabianHofmann** [comment]
Alright, I'm done so far. Three things small remaining points on the discussion list:

* rename argument `module` into `modules`?
* non-exploded Series (so Series of lists) for available_features and prepared_features? 
* `gc.collect` in prepare function due to possible file handling errors?

I'd say we can unworried discuss those things later after merging (as I'm already spending too much time with this, i'd be happy to wrap up).

File: atlite/datasets/era5.py [hunk 7]
```diff
@@ -187,8 +187,8 @@ def get_data_height(retrieval_params):
 
 def _area(coords):
     # North, West, South, East. Default: global
-    x0, x1 = coords['x'].min().item(), coords['x'].max().item()
-    y0, y1 = coords['y'].min().item(), coords['y'].max().item()
+    x0, x1 = coords.indexes['x'][[0, -1]]
+    y0, y1 = coords.indexes['y'][[0, -1]]
```
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply atlite/datasets/era5.py]: You're right, I reverted it

[the remark this answers is not in the document]
Comment (coroa) [reply atlite/datasets/era5.py]: Don't agree with the reasons.

- Min, max is not clearer than first and last item. 
- And you changed it in the first place to work around the bug that the convention was not followed yet in sarah which is now fixed for good.

File: atlite/cutout.py [hunk 8]
```diff
@@ -16,184 +16,207 @@
...
-                                          self.data.attrs['module']]
+
 
     @property
     def name(self):
         return self.path.stem
 
+    @property
+    def module(self):
+        return self.data.attrs.get('module')
```
[the remark this answers is not in the document]
Comment (coroa) [reply atlite/cutout.py:199]: You broke backwards compatibility already by renaming it from `dataset_module` to `module`. So the harm is already done!

File: atlite/convert.py [hunk 9]
```diff
@@ -93,134 +93,122 @@ def convert_and_aggregate(cutout, convert_func, windows=None, matrix=None,
...
+            geoseries_like = (pd.Series, gpd.GeoDataFrame, gpd.GeoSeries)
+            if isinstance(shapes, geoseries_like) and index is None:
+                index = shapes.index
+                index.name = 'shapes' if index.name is None else index.name
+            matrix = cutout.indicatormatrix(shapes, shapes_proj)
+
+        if matrix is not None:
+            matrix = sp.sparse.csr_matrix(matrix)
+            if index is None:
+                index = pd.RangeIndex(matrix.shape[0], name='shapes')
```
[the remark this answers is not in the document]
Comment (coroa) [reply atlite/convert.py]: Fair!

Only note, that I proposed, `index = pd.Index(index, name="shapes")` so as to not mutate index itself, but create a new index instead. This intention is probably even better communicated by using:
```python
if index.name is None:
    index = index.rename('dim_0')
```

[the remark this answers is not in the document]
Comment (FabianHofmann) [reply atlite/convert.py]: alright sounds good! added in https://github.com/PyPSA/atlite/commit/f370c5ddde0494f2913e8b0d745b944116d22fb0

File: atlite/datasets/era5.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply atlite/datasets/era5.py]: I really don't care, one assumes a convention, the other is independent of it. Both work. Performance wise both is negligibly (100-300 mu secs).

File: atlite/cutout.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply atlite/cutout.py:199]: Disclaimer on this: The `module` argument is the same `module` argument as in previous versions. It's a string (or list of strings), not python module(s). 

https://github.com/PyPSA/atlite/blob/fd010b2283a12498218aaa5b56173f1c6e30ef45/atlite/cutout.py#L78

File: atlite/datasets/era5.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (coroa) [reply atlite/datasets/era5.py]: Don’t care enough for adding a commit.

File: atlite/cutout.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (coroa) [reply atlite/cutout.py:199]: Sure, that’s why I think the name change is warranted.

[the remark this answers is not in the document]
Comment (coroa) [reply atlite/cutout.py:199]: But I see your backward compatibility concern now. It makes sense to keep meta.attrs[‘module’] and it makes sense to reuse this name, when exposing it on the cutout.

**Pull request merged**